contracts/utils/types/Time.sol

```diff
@@ -0,0 +1,176 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.19;
+
+library Time {
```

Comment:
Yes, UDVT adds verbosity. and it is a downside. (honestly its not that bad, its not worst than pre 0.8.0 math).
On the other hand I think the type safety is a real plus. If prevents you from adding two timestamps.

The main issue we had with Timers (IMO) was the fact they were deeply connected to block.timestamp, and we could not use them for block numbers. Here, its mostly about adding timepoints and duration. The presence of `block.timestamp` is really reduced to a minimum. Apart from `isPast` everything is block.number compatible. We could have a `isBefore(Timepoint t1, Timepoint t2)` for that.

---

TLDR:
- Yes there are some Timers vibe, but I designed that so that issues we add with Timers would not apply here
- Its compatible with block.number (with a single exception that can easily be replaced)
- Type checking is awesome
- Verbosity is not that bad IMO

Reply:
In particular the `type Delay is uint112;` is an abstraction that we built in a non reusable way for ACDAR. This is something that we really need accessible with an interface that abstracts the complexity. That is not something you replace with `schedule + setback < block.timestamp`

Reply:
Catching up here, I agree it's verbose and difficult to follow but I think this particular abstraction makes sense, especially for preserving the guarantee of a delay when it's reduced.

My understanding of that guarantee is that we have a strong opinion of how to deal with an update, and thus such opinion should be reflected as a recommendation in the library as a whole (the repository, not this `library` itself).

> Instead of timepoint.add(setback).isPast(), it's a lot more familiar and intuitive to see schedule + setback < block.timestamp.

Yes because we've seen it through multiple reviews. I won't expect a regular developer to be this familiar with the delay's delay. Also, note that the subtle difference between `<` and `<=` is error-prone and potentially dangerous.

I'd try to find a way to abstract our "recommendation" for delay updates but I'm still not sure if it's worth making a `Time` library at the moment. Perhaps UDVT with structs would reduce verbosity and make things clearer but it's not yet supported 😞

Reply:
@ernestognw for the record, [this is what it used to look like with `Timepoint` and a `Duration` UDVT ](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/308df371443910e15678435027f59a5fdd7e13c1/contracts/utils/types/Time.sol)

Reply:
Yes, this comment was about a previous version of the library a lot of which was cut out. Resolving.